networking/kvm.go

```diff
@@ -75,7 +89,100 @@ func kvmSetupNetAddressing(network *Networking, n activeNet, ifName string) erro
 		return fmt.Errorf("net-plugin returned no IPv4 configuration")
 	}
 
-	n.runtime.IP, n.runtime.Mask, n.runtime.HostIP = result.IP4.IP.IP, net.IP(result.IP4.IP.Mask), result.IP4.Gateway
+	n.runtime.IP, n.runtime.Mask, n.runtime.HostIP, n.runtime.IP4 = result.IP4.IP.IP, net.IP(result.IP4.IP.Mask), result.IP4.Gateway, result.IP4
+	return nil
+}
+
+func ensureHasAddr(link netlink.Link, ipn *net.IPNet) error {
+	addrs, err := netlink.AddrList(link, syscall.AF_INET)
+	if err != nil && err != syscall.ENOENT {
+		return fmt.Errorf("could not get list of IP addresses: %v", err)
+	}
+
+	// if there're no addresses on the interface, it's ok -- we'll add one
+	if len(addrs) > 0 {
+		ipnStr := ipn.String()
+		for _, a := range addrs {
+			// string comp is actually easiest for doing IPNet comps
+			if a.IPNet.String() == ipnStr {
+				return nil
+			}
+		}
+		return fmt.Errorf("%q already has an IP address different from %v", link.Attrs().Name, ipn.String())
+	}
+
+	addr := &netlink.Addr{IPNet: ipn, Label: link.Attrs().Name}
+	if err := netlink.AddrAdd(link, addr); err != nil {
+		return fmt.Errorf("could not add IP address to %q: %v", link.Attrs().Name, err)
+	}
+	return nil
+}
+
+func bridgeByName(name string) (*netlink.Bridge, error) {
+	l, err := netlink.LinkByName(name)
+	if err != nil {
+		return nil, fmt.Errorf("could not lookup %q: %v", name, err)
+	}
+	br, ok := l.(*netlink.Bridge)
+	if !ok {
+		return nil, fmt.Errorf("%q already exists but is not a bridge", name)
+	}
+	return br, nil
+}
+
+func ensureBridgeIsUp(brName string, mtu int) (*netlink.Bridge, error) {
+	br := &netlink.Bridge{
+		LinkAttrs: netlink.LinkAttrs{
+			Name: brName,
+			MTU:  mtu,
+		},
+	}
+
+	if err := netlink.LinkAdd(br); err != nil {
```

Comment:
logic looks a little twisted here.
your code says: "try adding a bridge, then if for "exist error", you try to find it and the try it up"

most of the time, the code, should just check the bridge exists and is up, so I prefer [LBYL](https://docs.python.org/2/glossary.html#term-lbyl) over [EAFP](https://docs.python.org/2/glossary.html#term-eafp) here

why not:

1. bridgeByName
    1. if not exists - create
    2. if exists - make sure match our requirements (that is our bridge, has required properties)
2. if not up, bring it up